### ISSUE TITLE
strictARP Configuration in kube-proxy (IPVS Mode) Does Not Revert sysctl Parameters on Setting to false

### ISSUE THREAD

**Issue by magicsong** [kubernetes/kubernetes#123992]
### What happened?

When `strictARP` is set to `true` in kube-proxy's configuration for IPVS mode, kube-proxy modifies certain `sysctl` parameters as expected. However, if the `strictARP` is later changed to `false`, the modified `sysctl` parameters are not reverted to their original values.

Upon inspecting the source code, I noticed that there is logic to apply changes when `strictARP` is `true`, but there appears to be no corresponding logic to revert these changes when `strictARP` is set to `false`.

### What did you expect to happen?

revert

### How can we reproduce it (as minimally and precisely as possible)?

master branch

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by magicsong** [comment]
@lizhipeng629

**Comment by cyclinder** [comment]
well, I think this should be a point that can be improved. This sysctl is a host-level configuration, and we should avoid it affecting the entire host.

/cc @uablrek @aojea

**Comment by uablrek** [comment]
IMO `kube-proxy` shouldn't tamper with sysctl at all, and never when it isn't ansolutely necessary. Someone may want the sysctl set for other reasons, and `kube-proxy` should not reset it.

**Comment by cyclinder** [comment]
When kube-proxy is in ipvs mode and using metallb as a load balancing service, it should be required to configure `strictARP`. see https://metallb.universe.tf/installation/#preparation.

If strictARP is set to false or kube-proxy mode switches to iptables mode, should we restore the sysctl? I think yes, since we no longer need to configure it and we don't want it to affect the rest of the host's progress.

**Comment by uablrek** [comment]
You can't say "restore", since you don't know if it was `kube-proxy` that set it in the first place. It may be set by a user, e.g. in `/etc/sysctl.conf`, because they need it.

**Comment by cyclinder** [comment]
> You can't say "restore"

sorry for my bad words. I would say that if the user needs it then they would set `strictARP ` to true and then kube-proxy sets the sysctl. If the user still needs it, I don't think they would want to change `strictARP` to false?

If they really don't need it anymore, currently, they need to set the sysctl manually for each node, so it would be nice if we could set `strictARP` to false and then kube-proxy could help us set the sysctl.

**Comment by magicsong** [comment]
yes, this is why the issues come up here. i just thought kube-proxy would rollback the modification.

> > You can't say "restore"
> 
> sorry for my bad words. I would say that if the user needs it then they would set `strictARP ` to true and then kube-proxy sets the sysctl. If the user still needs it, I don't think they would want to change `strictARP` to false?
> 
> If they really don't need it anymore, currently, they need to set the sysctl manually for each node, so it would be nice if we could set `strictARP` to false and then kube-proxy could help us set the sysctl.

**Comment by uablrek** [comment]
The kernel doc for reference:
```
arp_ignore - INTEGER
	Define different modes for sending replies in response to
	received ARP requests that resolve local target IP addresses:

	- 0 - (default): reply for any local target IP address, configured
	  on any interface
	- 1 - reply only if the target IP address is local address
	  configured on the incoming interface
	- 2 - reply only if the target IP address is local address
	  configured on the incoming interface and both with the
	  sender's IP address are part from same subnet on this interface
	- 3 - do not reply for local addresses configured with scope host,
	  only resolutions for global and link addresses are replied
	- 4-7 - reserved
	- 8 - do not reply for all local addresses

	The max value from conf/{all,interface}/arp_ignore is used
	when ARP request is received on the {interface}
```
When "strictARP" it true, `kube-proxy` ensures arp_ignore=1. In K8s this is AFAIK _only_ needed for proxy-mode=ipvs _and_ MetalLB in `L2 mode`.

For other reasons the cluster admin may however _want_ `arp_ignore != 0` (not necessarily 1) and explicitly configure it on node start-up. `Kube-proxy` MUST NOT force it to `0`!

`StrictARP=true` is never _required_ for `kube-proxy` in iptables or nftables mode (even though the cluster admin may want `arp_ignore != 0` for other reasons, as mentioned). So, I propose that `strictARP` is deprecated and obsoleted when the nftables mode becomes default.

**Comment by magicsong** [comment]
> For other reasons the cluster admin may however want arp_ignore != 0 (not necessarily 1) and explicitly configure it on node start-up. Kube-proxy MUST NOT force it to 0!
this logic not make sense. see below: 
"also  Kube-proxy MUST NOT force it to !0  if cluster admin set it to 0"

**Comment by aojea** [comment]
This was discussed today during sig network meeting, an important context is that we have a debt we need to deal with that is setting sysctl and that we are trying to avoid keep doing it.

We have now an option in kube-proxy IPVS that says 

> strictARP configures arp_ignore and arp_announce to avoid answering ARP queries from kube-ipvs0 interface

this effectively means that if strictARP:
- is false, it does not do anything
- is true kube-proxy set `net/ipv4/conf/all/arp_ignore` to 1 and `net/ipv4/conf/all/arp_announce` to 2

https://github.com/kubernetes/kubernetes/blob/f4e246bc93ffb68b33ed67c7896c379efa4207e7/pkg/proxy/ipvs/proxier.go#L325-L335

The request seems to be to give meaning to strictARP equal false to mean set `net/ipv4/conf/all/arp_ignore` to 0 and `net/ipv4/conf/all/arp_announce` to 0

The reason we can not do this is because as Lars says, if we make false to set some specific values, we'll always been overriding the sysctl config, since this option is a bool there is no way to differentiate if the user is indicating that it wants to `restore strictARP`  values or just is using the default value that means "leave the sysctl as they are"


https://github.com/kubernetes/kubernetes/blob/f4e246bc93ffb68b33ed67c7896c379efa4207e7/pkg/proxy/apis/config/types.go#L70-L72


/close
